File: Frontend/src/services/upvoteService.ts
import api from './api';

export const upvoteService = {
    // Add upvote
    addUpvote: async (userId: number, issueId: number): Promise<void> => {
        await api.post(`/upvotes?userId=${userId}&issueId=${issueId}`);
    },

    // Remove upvote
    removeUpvote: async (userId: number, issueId: number): Promise<void> => {
        await api.delete(`/upvotes?userId=${userId}&issueId=${issueId}`);
    },

    // Check if user has upvoted
    checkUpvote: async (userId: number, issueId: number): Promise<boolean> => {
        const response = await api.get(`/upvotes/check?userId=${userId}&issueId=${issueId}`);
        return response.data.hasUpvoted;
    },

    // Get upvote count for an issue
    getUpvoteCount: async (issueId: number): Promise<number> => {
        const response = await api.get(`/upvotes/count/${issueId}`);
        return response.data.count;
    },
};

File: Backend/src/main/java/com/elytra/backend/Services/UpvoteService.java
package com.elytra.backend.Services;

import com.elytra.backend.Models.Issue;
import com.elytra.backend.Models.Upvote;
import com.elytra.backend.Models.User;
import com.elytra.backend.Repository.IssueRepository;
import com.elytra.backend.Repository.UpvoteRepository;
import com.elytra.backend.Repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
@SuppressWarnings("null")
public class UpvoteService {

    @Autowired
    private UpvoteRepository upvoteRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private IssueRepository issueRepository;

    public Upvote addUpvote(Long userId, Long issueId) {
        // Check if upvote already exists
        if (upvoteRepository.existsByUserIdAndIssueId(userId, issueId)) {
            throw new RuntimeException("User has already upvoted this issue");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));

        Issue issue = issueRepository.findById(issueId)
                .orElseThrow(() -> new RuntimeException("Issue not found with id: " + issueId));

        Upvote upvote = new Upvote();
        upvote.setUser(user);
        upvote.setIssue(issue);

        // Increment upvote count on issue
        issue.setUpvotes(issue.getUpvotes() + 1);
        issueRepository.save(issue);

        return upvoteRepository.save(upvote);
    }

    public void removeUpvote(Long userId, Long issueId) {
        if (!upvoteRepository.existsByUserIdAndIssueId(userId, issueId)) {
            throw new RuntimeException("Upvote not found");
        }

        Issue issue = issueRepository.findById(issueId)
                .orElseThrow(() -> new RuntimeException("Issue not found with id: " + issueId));

        // Decrement upvote count on issue
        issue.setUpvotes(Math.max(0, issue.getUpvotes() - 1));
        issueRepository.save(issue);

        upvoteRepository.deleteByUserIdAndIssueId(userId, issueId);
    }

    public boolean hasUserUpvoted(Long userId, Long issueId) {
        return upvoteRepository.existsByUserIdAndIssueId(userId, issueId);
    }

    public Long getUpvoteCount(Long issueId) {
        return upvoteRepository.countByIssueId(issueId);
    }
}
